The DREGS class

stuartmcalpine opened the pull request:
Bit of a refactoring in this.

Essentially, there is now a top level DREGS class that handles all the machinery of the dataregistry. This is all there user has to import, and deals with connecting to the database, registering, and querying.

If the user has the default setup, the setup is reduced to

```
from dataregistry import DREGS

dregs = DREGS()

dregs.Registrar.register_dataset(...)
```

Docs have also been update to reflect this.

It also doesnt stop anyone doing it the original way, but this removes lots of overhead

JoanneBogart commented on src/dataregistry/registrar.py:
```diff
@@ -170,9 +183,16 @@ def _handle_data(self, relative_path, old_location, verbose):
 
         return dataset_organization, num_files, total_size
 
-    _MAX_CONFIG = 10000
```
Suggest keeping this line, except change `_MAX_CONFIG` to `_DEFAULT_MAX_CONFIG`.
then for register_execution change default value of `max_config_length` to `_DEFAULT_MAX_CONFIG`

stuartmcalpine replied:
Done

JoanneBogart commented on src/dataregistry/registrar.py:
```diff
@@ -270,41 +306,63 @@ def register_dataset(self, relative_path, version,
-        verbose : bool
+        verbose : bool, optional
             Provide some additional output information
+        owner : str, optional
```
These arguments won't be needed. Values will come from the Registrar object when it's initialized.

JoanneBogart commented on src/dataregistry/registrar.py:
```diff
@@ -270,41 +306,63 @@ def register_dataset(self, relative_path, version,
         """
 
-        if (self._owner_type == "production"):
+        # Make sure the owner_type is legal
```
All the code having to do with setting and verifying values for owner and owner_type should be moved to ```Registrar.__init__```

JoanneBogart reviewed (changes requested):
In summary, there are two overall comments which I alluded to in multiple places.
1. owner and owner_type should be a property of the Registrar object so they don't have to be supplied each time a dataset is registered. This is the one I feel most strongly about.
2. might as well cache the userid in the Registrar object as well.
and one more:
3. gen_filter more naturally belongs to the Query class

stuartmcalpine commented on src/dataregistry/DREGS.py:
```diff
@@ -0,0 +1,80 @@
+            engine, dialect, schema_version=self.schema_version, root_dir=root_dir
+        )
+
+    def gen_filter(self, property_name, bin_op, value):
```
Moved over to Query class

stuartmcalpine commented:
@JoanneBogart  I have made the changes to 2. and 3, thanks for the comments.

About point 1., I felt it was more natural to have the have all the options together for registering the dataset, i.e, having owner and owner type under `register_dataset`. My thoughts were:

- You are not restricted to a fixed owner and owner_type. If you wanted to register two datasets with different groups you would have to create two DREGS classes
- The owner and owner type default to $USER and "user", so for the majority of cases the user would not have to set those fields anyway.
- Owner and ownertype would have to be set initializing the DREGS class, which didn't feel natural as it only effects registering datasets

JoanneBogart commented:
@stuartmcalpine 
Although it's true a caller is not restricted to a fixed owner and owner_type, in the likely  most common writer use case (and the one motivating the dataregistry package), they will be fixed.  We need to make the process streamlined for this use case.  For normal (not administrator) users, I think it quite unlikely they would combine writing to different spaces in the same job. 
I don't know that "user" _is_ appropriate for the majority of cases.  That's true while code is under initial development so it is a reasonable default (later "project" or "group" might be as common).  But there should be a way to specify something else - just once, not for every dataset registration
As for your third point - it's true these settings only affect writers, but that's the nature of the thing and part of the reason the Registrar and Query classes are separate.  There will be users who only want to query, some via the API and some using the CLI.  There will be users who want to exclusively register new entries and some number (my guess is a minority) who will want to do both in the same session. The register or register/query users, as soon as they get serious, will probably prefer the API. 
It would be ok with me to leave owner and owner_type as optional arguments to register_dataset if we also have them (as optional arguments) to init routines for Registrar and DREGS.   Behavior would be:
* if specified in register_dataset, that takes precedence
* else if specified when Registrar object is initialized, use those values
* if specified in neither place, default to user values.

stuartmcalpine commented:
That was my thoughts for a possible solution also.

I've added the options to the `Registrar` and `DREGS` class to set a global `owner` and `owner_type` for that instance. The ordering is as you suggested, whats in `register_dataset` will overwrite it.

I've also added a test in the CI for it.

JoanneBogart reviewed (approved):
This looks fine now.  Thanks.
